src/main/storm/mesos/MesosNimbus.java

```diff
@@ -114,94 +120,228 @@ public String getHostName(Map<String, SupervisorDetails> map, String nodeId) {
+      LOG.error("Failed to prepare scheduler ", e);
       throw new RuntimeException(e);
     }
   }
+
   @SuppressWarnings("unchecked")
   protected void initialize(Map conf, String localDir) throws Exception {
-      _conf  = conf;
-      _state = new LocalState(localDir);
-      _allowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_ALLOWED_HOSTS));
-      _disallowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_DISALLOWED_HOSTS));
-      _scheduler = new NimbusScheduler();
-      createLocalServerPort();
-      setupHttpServer();
+    _conf = new HashMap();
+    _conf.putAll(conf);
+
+    _state = new LocalState(localDir);
+    _allowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_ALLOWED_HOSTS));
+    _disallowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_DISALLOWED_HOSTS));
+    Boolean preferReservedResources = (Boolean) conf.get(CONF_MESOS_PREFER_RESERVED_RESOURCES);
+    if (preferReservedResources != null) {
+      _preferReservedResources = preferReservedResources;
+    }
+    _container = Optional.fromNullable((String) conf.get(CONF_MESOS_CONTAINER_DOCKER_IMAGE));
+    _scheduler = new NimbusScheduler(this);
+    createLocalServerPort();
+    setupHttpServer();
+
+    _conf.put(Config.NIMBUS_HOST, _configUrl.getHost());
+    // Generate YAML to be served up to clients
+    Path generatedConfPath = Paths.get("generated-conf");
+    if (!generatedConfPath.toFile().exists() && !generatedConfPath.toFile().mkdirs()) {
+      throw new RuntimeException("Couldn't greate generated-conf dir");
+    }
+    File generatedConf = Paths.get(generatedConfPath.toString(), "storm.yaml").toFile();
+    Yaml yaml = new Yaml();
+    FileWriter writer = new FileWriter(generatedConf);
+    yaml.dump(_conf, writer);
+  }
+
+
+  public void doRegistration(final SchedulerDriver driver, Protos.FrameworkID id) {
+    _driver = driver;
+    try {
+      _state.put(FRAMEWORK_ID, id.getValue());
+    } catch (IOException e) {
+      LOG.error("Halting process...", e);
+      Runtime.getRuntime().halt(1);
+    }
+    Number filterSeconds = Optional.fromNullable((Number) _conf.get(CONF_MESOS_OFFER_FILTER_SECONDS)).or(120);
+    final Protos.Filters filters = Protos.Filters.newBuilder()
+        .setRefuseSeconds(filterSeconds.intValue())
+        .build();
+    _offers = new RotatingMap<>(
+        new RotatingMap.ExpiredCallback<Protos.OfferID, Protos.Offer>() {
+          @Override
+          public void expire(Protos.OfferID key, Protos.Offer val) {
+            driver.declineOffer(
+                val.getId(),
+                filters
+            );
+          }
+        }
+    );
+
+    Number lruCacheSize = Optional.fromNullable((Number) _conf.get(CONF_MESOS_OFFER_LRU_CACHE_SIZE)).or(1000);
+    final int intLruCacheSize = lruCacheSize.intValue();
+    _usedOffers = Collections.synchronizedMap(new LinkedHashMap<Protos.TaskID, Protos.Offer>(intLruCacheSize + 1, .75F, true) {
```

Comment:
It's used for the purpose of reusing existing slots (or offers). In other words, if a worker was already launched with an offer, but we can now repurpose that slot (or offer), it will first kill the task then launch a new one.

To be honest, I'm not even sure if this feature works. If we're going to change that, let's do it as part of a separate PR.